src/System.Private.CoreLib/src/System/Threading/LowLevelMonitor.Unix.cs

```diff
@@ -0,0 +1,139 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using Internal.Runtime.Augments;
+
+namespace System.Threading
+{
+    /// <summary>
+    /// Wraps a non-recursive mutex and condition.
+    /// 
+    /// Used by the wait subsystem on Unix, so this class cannot have any dependencies on the wait subsystem.
+    /// </summary>
+    internal sealed class LowLevelMonitor : IDisposable
+    {
+        private IntPtr _nativeMonitor;
+
+#if DEBUG
+        private RuntimeThread _ownerThread;
+#endif
+
+        public LowLevelMonitor()
+        {
+            _nativeMonitor = Interop.Sys.LowLevelMonitor_New();
+            if (_nativeMonitor == IntPtr.Zero)
+            {
+                throw new OutOfMemoryException();
+            }
+
+#if DEBUG
+            _ownerThread = null;
+#endif
+        }
+
+        ~LowLevelMonitor()
+        {
+            Dispose();
+        }
+
+        public void Dispose()
+        {
+            VerifyIsNotLockedByAnyThread();
+
+            if (_nativeMonitor == IntPtr.Zero)
+            {
+                return;
+            }
+
+            Interop.Sys.LowLevelMonitor_Delete(_nativeMonitor);
+            _nativeMonitor = IntPtr.Zero;
+            GC.SuppressFinalize(this);
+        }
+
+        public void VerifyIsLocked()
```

Comment:
I tried that first, but it looks like it still parses the code in Release builds and tries to find variables which are declared under #if DEBUG. Had the same issue with Debug.Assert statements that have expressions using variables declared under #if DEBUG. I couldn't figure out a way to make it work so I just switched to #if DEBUG where it was problematic.